collector/report.go

```diff
@@ -0,0 +1,331 @@
+
+	// These are for collecting metrics
+	Platform        string `json:"platform"`
+	SoftwareName    string `json:"software_name"`
+	SoftwareVersion string `json:"software_version"`
+
+	Path        string
+	IsEmpty     bool
+	expiryTimer *time.Timer
+	mutex       sync.Mutex
+}
+
+func (a *ActiveReport) SetFromEntry(e *MeasurementEntry) error {
+	a.ProbeCC = e.ProbeCC
+	a.ProbeASN = e.ProbeASN
+	a.TestName = e.TestName
+	if e.Annotations != nil {
+		annotations := e.Annotations.(map[string]interface{})
+		platform, ok := annotations["platform"].(string)
+		if ok {
+			a.Platform = platform
+		}
+	}
+	if err := a.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Validate checks that we have a valid report submitted
+func (a *ActiveReport) Validate() error {
+	if testNameRegexp.MatchString(a.TestName) != true {
+		return errors.New("invalid \"test_name\" field")
+	}
+	if probeASNRegexp.MatchString(a.ProbeASN) != true {
+		return errors.New("invalid \"probe_asn\" field")
+	}
+	if probeCCRegexp.MatchString(a.ProbeCC) != true {
+		return errors.New("invalid \"probe_cc\" field")
+	}
+	if stringInSlice(a.Format, supportedFormats) != true {
+		return fmt.Errorf("invalid \"format\" field, %s", a.Format)
+	}
+	return nil
+}
+
+// IncomingFilename determines the filename of an incoming report
+func (a *ActiveReport) IncomingFilename() string {
+	if stringInSlice(a.Format, supportedFormats) != true {
+		// Defensive coding
+		panic("an unexpected value for format was found. Bailing...")
+	}
+	return fmt.Sprintf("%s.%s",
+		a.ReportID,
+		a.Format)
+}
+
+// SyncFilename creates a filename in the sync directory of the active report
+func (a *ActiveReport) SyncFilename() (string, error) {
+	// Defensive coding.
+	if err := a.Validate(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s-%s-%s-%s-probe-0.2.0.%s",
+		a.CreationTime.Format(TimestampFormat),
+		a.TestName,
+		a.ReportID,
+		a.ProbeASN,
+		a.ProbeCC,
+		a.Format), nil
+}
+
+// BackendExtra is serverside extra metadata
+type BackendExtra struct {
+	SubmissionTime time.Time `json:"submission_time" bson:"submission_time"`
+	MeasurementID  string    `json:"measurement_id" bson:"measurement_id"`
+	ReportID       string    `json:"report_id" bson:"report_id"`
+}
+
+// MeasurementEntry is the structure of measurements submitted by an OONI Probe client
+type MeasurementEntry struct {
+	// These values are added by the pipeline
+	// BucketDate     string `json:"bucket_date"`
+	// ReportFilename string `json:"report_filename"`
+
+	ID                   string       `json:"id"`
+	ReportID             string       `json:"report_id"`
+	TestName             string       `json:"test_name"`
+	TestVersion          string       `json:"test_version"`
+	MeasurementStartTime string       `json:"measurement_start_time,omitempty"`
+	TestStartTime        string       `json:"test_start_time"` // XXX these should actually be time
+	Annotations          interface{}  `json:"annotations"`
```

Comment:
Annotations are `map[string]string` in MK. What are they in OONI legacy?

Reply:
I believe we never really enforced anything as strict as that in OONI legacy, hence we are using the opaque `interface{}` in here for it.

Reply:
How about `map[string]interface{}` then?